### PR TITLE
CI: Transition from s4ext to json

### PR DESCRIPTION
**Pull request by jcfr**
In response to discussions during the Slicer developer hangout on March 12th, we have decided to revamp the organization of extension metadata in the extensions index, transitioning from the s4ext format to json.

Motivation:
* Eliminate redundant and unused information from the "description" file.
* Simplify programmatic parsing of the "description" files
* Decouple the metadata organized in the extension `CMakeLists.txt` from the ones organized in this repository and used to drive the build of extensions.
* Enable Slicer maintainers to define and update the extension category independently[^1] of the upstream extension sources
* Prepare for the integration of the upcoming "tier" metadata.[^2]

[^1]: The `EXTENSION_CATEGORY` variable currently set in the `CMakeLists.txt` will likely be deprecated

[^2]: This is described on the Slice Discourse forum: https://discourse.slicer.org/t/introduction-of-tiers-for-slicer-extensions/34870

Related issues:
* https://github.com/Slicer/Slicer/issues/6434

Related pull requests:
* https://github.com/Slicer/Slicer/pull/7629
* https://github.com/Slicer/Slicer/pull/7644

Next steps:
* [x] Add json schema
* [x] Update Extension build-system organized at https://github.com/Slicer/Slicer/tree/main/Extensions/CMake

cc: @lassoan @sjh26 @RafaelPalomar @mauigna06 @pieper

### CODE REVIEW

**Comment by jcfr** [comment]
For reference, the following script was used to convert from `s4ext` to `json`:

```python
import json
import sys
from pathlib import Path

extensions_index_dir = Path("/home/jcfr/Projects/ExtensionsIndex")
updated_extensions_index_dir = extensions_index_dir


def parse_s4ext(ext_file_path):
    """Parse a Slicer extension description file.
    :param ext_file_path: Path to a Slicer extension description file.
    """
    ext_metadata = {}
    with open(ext_file_path) as ext_file:
        for line in ext_file:
            if not line.strip() or line.startswith("#"):
                continue
            fields = [field.strip() for field in line.split(' ', 1)]
            assert(len(fields) <= 2)
            ext_metadata[fields[0]] = fields[1] if len(fields) == 2 else None
    return ext_metadata


# Collect s4ext files
s4ext_filepaths = list(extensions_index_dir.glob("*.s4ext"))

print(f"Found {len(s4ext_filepaths)} extension files (.s4ext)")

# Parse s4ext files and generate corresponding json files
for index, filepath in enumerate(s4ext_filepaths):

    metadata = parse_s4ext(filepath)

    updated_metadata = {
        "scmurl": metadata["scmurl"],
        "scmrevision": metadata["scmrevision"],
        "depends": [] if metadata.get("depends", "NA") == "NA" else metadata["depends"].split(" "),
        "category": metadata["category"],
    }

    with open(updated_extensions_index_dir / f"{filepath.stem}.json", 'w') as fileContents:
        fileContents.write(json.dumps(updated_metadata, sort_keys=True, indent=4))
        fileContents.write("\n")


print(f"Generated {index + 1} extension files (.json)")

from pprint import pprint as pp

print(f"\nMetadata of extension #{index + 1} ({filepath.stem}):\n")
pp(updated_metadata)
```

**Comment by jcfr** [comment]
@lassoan  This is ready for review

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -28,5 +28,12 @@ repos:
   - repo: https://github.com/python-jsonschema/check-jsonschema
     rev: "0.28.0"
     hooks:
+      - id: check-jsonschema
+        files: ^[^/]*\.json$
+        args:
+          [
+            "--schemafile",
+            "schemas/slicer-extension-catalog-entry-schema-v1.0.0.json",
```
Comment (jcfr) [inline .pre-commit-config.yaml:36]: Once the schema will be integrated into Slicer repo, I will switch to using `check-metaschema` able to download the schema based on `$schema` property.

See https://check-jsonschema.readthedocs.io/en/stable/precommit_usage.html#pre-commit-usage

**Review by lassoan — COMMENTED** [review]
Thank you, it looks good to me.

I don't really like the inconsistent naming convention (`build_dependencies` uses `_` separator, while `scmrevision` does not use separator), but if you have a good reason for this then I can live with it.

**Comment by jcfr** [comment]
### Extension Updates

List of pull requests intended to fix and consolidate extension metadata:

* https://github.com/naterex23/SlicerAblationPlanner/pull/3

File: ABLTemporalBoneSegmentation.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "$schema": "https://raw.githubusercontent.com/Slicer/Slicer/main/Schemas/slicer-extension-catalog-item-schema-v1.0.0.json#",
```
Comment (jamesobutler) [inline ABLTemporalBoneSegmentation.json:2]: @jcfr Is this link needing to be changed?
```suggestion
  "$schema": "https://raw.githubusercontent.com/Slicer/Slicer/main/Schemas/slicer-extension-catalog-entry-schema-v1.0.0.json#",
```

Comment (jcfr) [reply ABLTemporalBoneSegmentation.json:2]: Good catch

Comment (jamesobutler) [reply ABLTemporalBoneSegmentation.json:2]: Same issue at https://github.com/Slicer/Slicer/blob/1b9bd54b06fa1b237a0caa8399c0a50de5b19283/Schemas/slicer-extension-catalog-entry-schema-v1.0.0.json#L3

Comment (jcfr) [reply ABLTemporalBoneSegmentation.json:2]: See https://github.com/Slicer/Slicer/pull/7710

**Pull request merged**